src/main/java/org/apache/datasketches/kll/KllDirectCompactItemsSketch.java

```diff
@@ -0,0 +1,307 @@
+
+  /**
+   * Internal implementation of the wrapped Memory KllSketch.
+   * @param memVal the MemoryValadate object
+   * @param comparator to compare items
+   * @param serDe Serializer / deserializer for items of type <i>T</i> and <i>T[]</i>.
+   */
+  KllDirectCompactItemsSketch( //called below and KllItemsSketch
+      final KllMemoryValidate memVal,
+      final Comparator<? super T> comparator,
+      final ArrayOfItemsSerDe<T> serDe) {
+    super(memVal.sketchStructure, comparator, serDe);
+    this.mem = memVal.srcMem;
+    readOnly = true;
+    levelsArr = memVal.levelsArr; //always converted to writable form.
+  }
+
+  @Override
+  public int getK() {
+    return getMemoryK(mem);
+  }
+
+  @Override
+  public T getMaxItem() {
+    if (sketchStructure == COMPACT_EMPTY || isEmpty()) {
+      throw new SketchesArgumentException(EMPTY_MSG);
+    }
+    if (sketchStructure == COMPACT_SINGLE) {
+      return serDe.deserializeFromMemory(mem, DATA_START_ADR_SINGLE_ITEM, 1)[0];
+    }
+    //sketchStructure == COMPACT_FULL
+    final int offset = DATA_START_ADR + getNumLevels() * Integer.BYTES;
+    return serDe.deserializeFromMemory(mem, offset, 2)[1];
+  }
+
+  @Override
+  public T getMinItem() {
+    if (sketchStructure == COMPACT_EMPTY || isEmpty()) {
+      throw new SketchesArgumentException(EMPTY_MSG);
+    }
+    if (sketchStructure == COMPACT_SINGLE) {
+      return serDe.deserializeFromMemory(mem, DATA_START_ADR_SINGLE_ITEM, 1)[0];
+    }
+    //sketchStructure == COMPACT_FULL
+    final int offset = DATA_START_ADR + getNumLevels() * Integer.BYTES;
+    return serDe.deserializeFromMemory(mem, offset, 1)[0];
+  }
+
+  @Override
+  public long getN() {
+    if (sketchStructure == COMPACT_EMPTY) { return 0; }
+    if (sketchStructure == COMPACT_SINGLE) { return 1; }
+    return getMemoryN(mem);
+  }
+
+  //restricted
+
+  private int getCompactDataOffset() { //Sketch cannot be empty
+    return sketchStructure == COMPACT_SINGLE
+        ? DATA_START_ADR_SINGLE_ITEM
+        : DATA_START_ADR + getNumLevels() * Integer.BYTES + getMinMaxSizeBytes();
+  }
+
+  @Override
+  int getM() {
+    return getMemoryM(mem);
+  }
+
+  @Override
+  MemoryRequestServer getMemoryRequestServer() {
+    return null; // not used. Must return null.
+  }
+
+  @Override
+  int getMinK() {
+    if (sketchStructure == COMPACT_EMPTY || sketchStructure == COMPACT_SINGLE) { return getMemoryK(mem); }
+    return getMemoryMinK(mem);
+  }
+
+  @Override
+  byte[] getMinMaxByteArr() {
+    if (sketchStructure == COMPACT_EMPTY || getN() == 0) { return new byte[0]; }
+    if (sketchStructure == COMPACT_SINGLE) {
+      final int bytesSingle = serDe.sizeOf(mem, DATA_START_ADR_SINGLE_ITEM, 1);
+      final byte[] byteArr = new byte[2 * bytesSingle];
+      mem.getByteArray(DATA_START_ADR_SINGLE_ITEM, byteArr, 0, bytesSingle);
+      mem.getByteArray(DATA_START_ADR_SINGLE_ITEM, byteArr, bytesSingle, bytesSingle);
+    }
+    //sketchStructure == COMPACT_FULL
```

Comment:
seems like COMPACT_SINGLE also reaches this point

Reply:
Very good catch!  Thank you! Fixed.